packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp

```diff
@@ -392,12 +383,12 @@ namespace Amesos2 {
           const size_t lclNumRows = redist_mv.getLocalLength();
           for (size_t j = 0; j < redist_mv.getNumVectors(); ++j) {
             auto av_j = new_data(lda*j, lclNumRows);
-            auto X_j = redist_mv.getVector(j);
             auto X_lcl_j_2d = redist_mv.template getLocalView<host_execution_space> ();
             auto X_lcl_j_1d = Kokkos::subview (X_lcl_j_2d, Kokkos::ALL (), j);
-            for ( size_t i = 0; i < lclNumRows; ++i ) {
-              X_lcl_j_1d(i) = av_j[i];
-            }
+
+            using val_type = typename decltype( X_lcl_j_1d )::value_type;
+            Kokkos::View<val_type*, Kokkos::HostSpace> umavj ( const_cast< val_type* > ( reinterpret_cast<const val_type*> ( av_j.getRawPtr () ) ), av_j.size () );
+            Kokkos::deep_copy (umavj, X_lcl_j_1d);
```

Comment:
1. This code begs for a refactor -- this function ("copy from an ArrayView to a MultiVector") seems to come up a few times.
2. Does `new_data` come from a MultiVector?  If so, you could just `Tpetra::deep_copy` between the two MultiVectors.

Reply:
@mhoemmen: "this function ("copy from an ArrayView to a MultiVector") seems to come up a few times"

Yeah, we've reimplemented some optimizations from Amesos to avoid this for single process cases when the data pointer can be used directly without worrying about collectives to bring all data to the root proc. Adding @srajama1, we can discuss some possible options, estimate amount of work, availability etc.

"Does `new_data` come from a MultiVector"
No, `new_data` is an ArrayView storing the result (on root proc) of the solve. It needs to be copied back and redistributed to the MultiVector passed in through the `solve(...)` call.

Reply:
@ndellingwood : Can you file a different issue for optimizing this code. Let us get the PR with the current changes pushed in and we can look at optimizing this as a next step.

Reply:
@srajama1 sounds good :)